### ISSUE TITLE
[ML] Boosted tree incremental training

### ISSUE THREAD

**Issue by tveasey** [elastic/ml-cpp#1721]
This covers the work selectively rip up and regrow and/or recompute leaf values for a decision tree for fine tuning. It should be possible to treat this as just another round of boosting. The work breaks down into the following tasks:
- [x] Leaf node statistics should be hierarchy with concrete implementations from train from scratch and incremental training
- [x] New leaf node statistics for incremental training
- [x] New loss function for MSE for incremental training (other loss functions will be the subject of a separate issue)
- [x] Select trees to retrain based on how much they can reduce improve loss
- [x] Retrain functionality for selected trees 
- [x] Cross-validate additional hyperparameters targeting prediction stability
- [x] Update interfaces to support incremental training
